File: C++/InterviewExam/2016/code/binary_search.cpp
/**
	题目描述

		对于一个有序数组，我们通常采用二分查找的方式来定位某一元素，请编写二分查找的算法，在数组中查找指定元素。
		给定一个整数数组A及它的大小n，同时给定要查找的元素val，请返回它在数组中的位置(从0开始)，若不存在该元素，返回-1。若该元素出现多次，请返回第一次出现的位置。
		测试样例：
		[1,3,5,7,9],5,3
		返回：1
*/

class BinarySearch {
public:
    int getPos(vector<int> A, int n, int val) {
        // write code here
        int lo = 0;
        int hi = n;
        while (lo < hi){
            int mid = (hi - lo) / 2 + lo;
            if (A[mid] < val){
                lo = mid + 1;
            }
            else{
                hi = mid;
            }
        }
        return (A[lo] == val) ? lo : -1;
    }
};

File: Java/CrackingTheCodingInterview/sort/binarysearch.java
/**
	二分查找
*/
int binarysearch(int[] array, int x){
	int low = 0;
	int high = array.length - 1;
	int mid;
	
	while (low <= high){
		mid = (low + high) / 2;
		if (array[low] < x){
			low = mid + 1;
		}
		else if (array[mid] > x){
			high = mid - 1;
		}
		else{
			return mid;
		}
	}
	
	return -1;
}

// 递归版
int binarysearchRecursive(int[] array, int x, int low, int high){
	if (low > high){
		return -1;
	}
	else if (low == high){
		return array[low] == x ? low : -1;
	}
	
	int mid = (low + high) / 2;
	if (array[mid] < x){
		binarysearchRecursive(array, x, mid + 1, high);
	}
	else if (array[mid] > x){
		binarysearchRecursive(array, x, low, mid - 1);
	}
	else{
		return mid;
	}
}